crates/wasmparser/src/validator/component_types.rs

```diff
@@ -34,56 +35,174 @@ const MAX_FLAT_FUNC_RESULTS: usize = 1;
 /// The maximum lowered types, including a possible type for a return pointer parameter.
 const MAX_LOWERED_TYPES: usize = MAX_FLAT_FUNC_PARAMS + 1;
 
+/// Error used when we are in the process of lowering a component signature to a
+/// core signature and we need to keep retry lowering in a different mode
+/// (e.g. flat vs indirect in linear memory) only for certain kinds of errors.
+enum LowerError {
+    OutOfCapacity,
+    TypeMismatch(BinaryReaderError),
+    Other(BinaryReaderError),
+}
+
+impl From<BinaryReaderError> for LowerError {
+    fn from(v: BinaryReaderError) -> Self {
+        Self::Other(v)
+    }
+}
+
+impl LowerError {
+    #[track_caller]
+    fn unwrap_binary_reader_error(self) -> BinaryReaderError {
+        match self {
+            LowerError::TypeMismatch(e) | LowerError::Other(e) => e,
+            LowerError::OutOfCapacity => panic!(),
+        }
+    }
+}
+
+type LowerResult<T = ()> = Result<T, LowerError>;
+
+enum Variance {
+    Covariant,
+    Contravariant,
+}
+
+/// A list of expected types and how to validate them.
+struct ExpectedLoweredTypes<'a> {
+    types: &'a TypeList,
```

Comment:
This stores a `TypeList` but elsewhere when we're passing around a `LoweredTypes` we also pass around a `TypeList`. Could this change to not storing a `TypeList` and taking a parameter, or instead storing this within `LoweredTypes` to uniformly either use-a-field or use-a-parameter?

Reply:
The thing is that we don't always have an `ExpectedLoweredTypes`, so we could add a `TypeList` to `LoweredTypes` or something to remove the threading, but that is additional plumbing work.

Alternatively, we could remove it from this type and pass it at every call site. But that seems kind of like unnecessary busy work as well, and also won't remove the lifetime parameter if we make `ExpectedTypes` hold a `&'a [ValType]` instead of a `SmallVec`.